### DISCUSSION TITLE
这里计算NULL参数的方式有BUG吧？

### DISCUSSION THREAD

**Discussion by aceyin** [cloudwu/skynet#1505] [General]
https://github.com/cloudwu/skynet/blob/350248a7cc1102356584ef6f2a6c602d8d42c52c/lualib/skynet/db/mysql.lua#L581

例如: args = {1,2,nil,4,nil,6}
按照代码里面的计算方式: null_count  = (#args + 7) // 8 得出的 null_count 是 1 
但明显， args 里面有2个nil

**Comment by cloudwu** [comment]
这里支持 nil  ?

**Comment by cloudwu** [comment]
似乎是后来支持的。 https://github.com/cloudwu/skynet/pull/1228

@zero-rp

**Comment by cloudwu** [comment]
看看 https://github.com/cloudwu/skynet/commit/1e2ed7e6ff7d86d7cc6395933e2f327a403a8e29 这样改？

**Comment by zhuilang** [comment]
mysql 950行貌似空跑了 @cloudwu

**Comment by cloudwu** [comment]
什么叫空跑。

**Reply by zhuilang** [reply]
没有用到返回值

**Reply by cloudwu** [reply]
我不觉得有什么问题。或者你可以去源头提 issue https://github.com/openresty/lua-resty-mysql
然后再同步过来。

**Reply by zhuilang** [reply]
local function read_execute_result(self, sock)
    local packet, typ, err = _recv_packet(self, sock)
    if not packet then
        return nil, err
        --error( err )
    end

    if typ == "ERR" then
        local errno, msg, sqlstate = _parse_err_packet(packet)
        return nil, msg, errno, sqlstate
        --error( strformat("errno:%d, msg:%s,sqlstate:%s",errno,msg,sqlstate))
    end

    if typ == "OK" then
        local res = _parse_ok_packet(packet)
        if res and res.server_status & SERVER_MORE_RESULTS_EXISTS ~= 0 then
            return res, "again"
        end
        return res
    end

    if typ ~= "DATA" then
        return nil, "packet type " .. typ .. " not supported"
        --error( "packet type " .. typ .. " not supported" )
    end

    -- typ == 'DATA'

    local field_count, extra = _parse_result_set_header_packet(packet)

    local cols = {}
    local col
    while true do
        packet, typ, err = _recv_packet(self, sock)
        if typ == "EOF" then
            local warning_count, status_flags = _parse_eof_packet(packet)
            break
        end
        col = _parse_field_packet(packet)
        if not col then
            break
            --error( strformat("errno:%d, msg:%s,sqlstate:%s",errno,msg,sqlstate))
        end
        table.insert(cols, col)
    end

    --没有记录集返回
    if #cols < 1 then
        return {}
    end

    local compact = self.compact
    local rows = {}
    local row
    while true do
        packet, typ, err = _recv_packet(self, sock)
        if typ == "EOF" then
            local warning_count, status_flags = _parse_eof_packet(packet)
            if status_flags & SERVER_MORE_RESULTS_EXISTS ~= 0 then
                return rows, "again"
            end
            break
        end
        row = _parse_row_data_binary(packet, cols, compact)
        if not col then
            break
        end
        table.insert(rows, row)
    end

    return rows
end

**Reply by zhuilang** [reply]
预处理是这边加的，
local field_count, extra = _parse_result_set_header_packet(packet)
用不到

**Reply by cloudwu** [reply]
没看明白，你可以提个 pr 。

**Reply by cloudwu** [reply]
我看明白了。但我觉得不是个问题。只要不是 bug , 我不想维护 mysql 的这段代码。如果有人想维护，可以承接此职责。
